docs/markdown/Kconfig-module.md

```diff
@@ -0,0 +1,52 @@
+---
+short-description: Unstable kconfig module
+authors:
+    - name: Mark Schulte, Paolo Bonzini
+      years: [2017, 2019]
+      has-copyright: false
+...
+
+# Unstable kconfig module
+
+This module parses Kconfig output files to allow use of kconfig
+configurations in meson projects.
+
+**Note**: this does not provide kconfig frontend tooling to generate a
+configuration. You still need something such as kconfig frontends (see
+link below) to parse your Kconfig files, and then (after you've
+choosen the configuration options), output a ".config" file.
+
+  [kconfig-frontends]: http://ymorin.is-a-geek.org/projects/kconfig-frontends
+
+## Usage
+
+The module may be imported as follows:
+
+``` meson
+kconfig = import('unstable-kconfig')
+```
+
+The following functions will then be available as methods on the object
+with the name `kconfig`. You can, of course, replace the name
+`kconfig` with anything else.
+
+### kconfig.load()
+
+This function loads a kconfig output file and returns a dictionary object.
+
+`kconfig.load()` makes no attempt at parsing the values in the
+file.  Therefore, true boolean values will be represented as the string "y"
+and integer values will have to be converted with `.to_int()`.
+
+Kconfig frontends usually have ".config" as the default name for the
+configuration file.  However, placing the configuration file in the source
+directory limits the user to one configuration per source directory.
+In order to allow separate configurations for each build directory, as is
+the Meson standard, `meson.build` should not hardcode ".config" as the
+argument to `kconfig.load()`, and should instead make the argument to
+`kconfig.load()` a [project build option](Build-options.md).
```

Comment:
Not hardcoding is great, unfortunately I just discovered the hard way a significant mismatch between  `kconfig.load()` and the way kconfig itself works. With meson/kconfig, you still need to "hardcode" in meson.build either  `kconfig.load(meson.build_root() / my_config)` or `kconfig.load(meson.source_root() / my_config)`. In the Linux kernel, the location of `.config` automatically changes when you add or remove the `O=` option and you don't need a different Makefile to build in versus out of source.

Worse: if you hardcode neither directory and just `kconfig.load(my_config)`, then meson will silently ignore any `my_config` file in `meson.build_root()` while `kconfiglib.py` (naturally) gives precedence to `build_root()`. So now part of the build may use one config and the rest another; that hurt.

Bug or feature? Should I file a new github issue about this @bonzini ?

Reply:
I think we should change this to say something like

    If you are using this with Kconfig, the `.config` file is placed by the Kconfig
    front-end in the build directory. Therefore, the argument to Kconfig should be
    `meson.build_root() / '.config'`.

Reply:
> the `.config` file is placed by the Kconfig front-end in the build directory.

This is not true with kconfiglib.py which conveniently places .config in the _current_ directory. This is convenient because interactive users can have one .config in the source tree that doesn't get deleted on `--wipe`, while automation can go the other way and easily build multiple configs from the same source tree. 

For the Linux kernel's I find it technically correct but quite confusing to name the source directory "build directory" for interactive users who don't use the 'O=` parameter.

Does QEMU's meson.build requires interactive users to create and maintain their .config in meson.build_root()? What about QEMU's Makefiles?